src/shell.rs

```diff
@@ -0,0 +1,70 @@
+use crate::cmd;
+use atty;
+use log::debug;
+use std::error::Error;
+use std::io::{stdout, BufRead, Stdin, StdinLock, Write};
+
+pub enum ShellInput {
+    Builtin(BuiltinCommands),
+    Command(cmd::Sub),
+    Eof,
+    ParseError(Box<dyn Error>),
+}
+
+pub enum BuiltinCommands {
+    Exit,
+}
+
+pub struct ShellReader<'a> {
+    line: String,
+    input: StdinLock<'a>,
+}
+
+impl<'a> ShellReader<'a> {
+    pub fn new(input: &'a Stdin) -> Self {
+        Self {
+            line: String::new(),
+            input: input.lock(),
+        }
+    }
+
+    pub fn read_line(&mut self) -> Result<ShellInput, Box<dyn Error>> {
+        if atty::is(atty::Stream::Stdout) {
+            print!("> ");
+            stdout().flush().expect("failed to flush output");
+        }
+
+        match self.input.read_line(&mut self.line) {
+            Ok(0) => {
+                // // append newline after '> '
+                // println!("");
+                return Ok(ShellInput::Eof);
+            }
+            Ok(_) => (),
+            Err(e) => return Err(Box::new(e)),
+        }
+
+        let line = self.line.trim_end();
+
+        debug!("Line read: {:?}", line);
+
+        match line {
+            // build-in shell command(s)
+            "exit" => Ok(ShellInput::Builtin(BuiltinCommands::Exit)),
+            // dy commands
+            line => {
+                // TODO: better handling of whitespaces
+                let args = line.split(' ');
```

Comment:
leaving it as naive. What syntax do you like (such as `"hoge fuga\npiyo"` or `'hoge fuga'\n'piyo'`)?

Reply:
I understand that current implementation requires a single whitespace as an argument separator. With regard to the multi-line syntax such as `"hoge fuga\npiyo"`, as the current dynein command is supposed to be passed on a single line, we do not have a specific policy on the syntax of passing arguments on multiple lines.

Possible syntaxes I can imagine would be (let me know if you have others promising options):
1. Accepts arguments with multiple lines and a command ends with semicolon(`;`) or other special character (like MySQL interactive shell).
2. Only when you put `\` at the end of line, the shell mode accepts next line (like `bash/zsh`).

I prefer No.2, but it's also ok to leave it as it is for now and work on it in future enhancements.

Reply:
I mentioned how to include whitespaces in *tokens* but you are saying how to write multiple-*line*  inputs.
Both are important, but should be discussed separately.

On how to write multiple lines, I'm for 2 too.

Concerning how to include whitespaces in tokens, consider this input:
`query --sort-key '= 12' pk`
This is parsed as `query`, `--sort-key`, `'=`, `12'`, `pk` in current implementation.
This is not intended behaviour. We need some quotation syntaxes. However, there are options which syntax to adopt.
`'= 12'` ? Or `"= 12"`? When it comes to include quotation marks, how to escape them?
The first example I showed wasn't sufficient to describe this option. I apologize for that.

Then, which syntax do you prefer? Or leave it as-is and discus the syntax in some new issues?
IMO multi-line syntax can be postponed but minimal implementation of quoting is mandatory.

Reply:
I appreciate you describing it in detail, sorry for my confusion. Now I understand your query and I agree with you that we need to make a minimum decision about the issue on whitespaces in tokens here.

I prefer single-quotes `'= 12'` than double-quotes `"= 12"`.
Then, if we choose single-quotes, we need to decide how to escape single-quotes in them. One naive idea is just escaping it with backquotes (e.g. `> get 'I\'m a hoge'`)... Do you think this method will work?

Reply:
Yes, and I'll implement it.
I have a question. What syntax do you like to escape `\` itself? I think `\\` would be natural. If it looks good to you, I'll implement it too.

Reply:
Yes, using `\\` to escape `\` itself sounds natural for me as well :)